vllm/v1/core/scheduler.py

```diff
@@ -219,12 +236,25 @@ def schedule(self) -> "SchedulerOutput":
 
         # Next, schedule the WAITING requests.
         if not preempted_reqs:
-            while self.waiting and token_budget > 0:
+            # NOTE: We uses num_to_skip to determine
+            # which guided request within the waiting queue to skip
+            # over if the FSM of said request are yet to be ready.
+            num_to_skip: int = 0
+            while num_to_skip < len(self.waiting) and token_budget > 0:
                 if len(self.running) == self.max_num_running_reqs:
                     break
 
-                request = self.waiting[0]
+                request = self.waiting[num_to_skip]
+
+                if (request.use_guided_decoding
+                        and request.status == RequestStatus.WAITING_FOR_FSM):
```

Comment:
Checking for `use_guided_decoding` doesn't seem necessary here

```suggestion
                if request.status == RequestStatus.WAITING_FOR_FSM:
```